### PR TITLE
feat: add retry/timeout to manual surface

### LINKED ISSUE #221: Expose retry, timeout parameters in manual surface
the generated surface surfaces retry and timeout. For many calls this would be useful for firestore customers.
```
def run_query(
        self,
        request: firestore.RunQueryRequest = None,
        *,
        retry: retries.Retry = gapic_v1.method.DEFAULT,
        timeout: float = None,
        metadata: Sequence[Tuple[str, str]] = (),
    ) -> Iterable[firestore.RunQueryResponse]:
```

Async and Sync clients should be updated to expose this.

### PR DESCRIPTION
**Pull request by tseaver**
## Base Classes
- [x] `base_client.BaseClient.collections` (signature only)
- [x] `base_client.BaseClient.get_all` (signature only)
- [x] `base_collection.BaseCollectionReference.add` (signature only)
- [x] `base_collection.BaseCollectionReference.get` (signature only)
- [x] `base_collection.BaseCollectionReference.list_documents` (signature only)
- [x] `base_collection.BaseCollectionReference.stream` (signature only)
- [x] `base_document.BaseDocumentReference.collections` (signature only)
- [x] `base_document.BaseDocumentReference.create` (signature only)
- [x] `base_document.BaseDocumentReference.delete` (signature only)
- [x] `base_document.BaseDocumentReference.get` (signature only)
- [x] `base_document.BaseDocumentReference.set` (signature only)
- [x] `base_document.BaseDocumentReference.update` (signature only)
- [x] `base_query.BaseCollectionGroup.get_partitions` (signature only)
- [x] `base_query.BaseQuery.get` (signature only)
- [x] `base_query.BaseQuery.stream` (signature only)
- [x] `base_transaction.BaseTransaction.get` (signature only)
- [x] `base_transaction.BaseTransaction.get_all` (signature only)

## Synchronous
- [x] `batch.Batch.commit`
- [x] `client.Client.collections`
- [x] `client.Client.get_all`
- [x] `collection.Collection.add`
- [x] `collection.Collection.get`
- [x] `collection.Collection.list_documents`
- [x] `collection.Collection.stream`
- [x] `document.Document.collections`
- [x] `document.Document.create`
- [x] `document.Document.delete`
- [x] `document.Document.get`
- [x] `document.Document.set`
- [x] `document.Document.update`
- [x] `query.CollectionGroup.get_partitions`
- [x] `query.Query.get`
- [x] `query.Query.stream`
- [x] `transaction.Transaction.get`
- [x] `transaction.Transaction.get_all`

## Asynchronous
- [x] `async_batch.AsyncBatch.commit`
- [x] `async_client.AsyncClient.collections`
- [x] `async_client.AsyncClient.get_all`
- [x] `async_collection.AsyncCollection.add`
- [x] `async_collection.AsyncCollection.get`
- [x] `async_collection.AsyncCollection.list_documents`
- [x] `async_collection.AsyncCollection.stream`
- [x] `async_document.AsyncDocument.collections`
- [x] `async_document.AsyncDocument.create`
- [x] `async_document.AsyncDocument.delete`
- [x] `async_document.AsyncDocument.get`
- [x] `async_document.AsyncDocument.set`
- [x] `async_document.AsyncDocument.update`
- [x] `async_query.AsyncCollectionGroup.get_partitions`
- [x] `async_query.AsyncQuery.get`
- [x] `async_query.AsyncQuery.stream`
- [x] `async_transaction.AsyncTransaction.get`
- [x] `async_transaction.AsyncTransaction.get_all`

## Unsure
- `base_collection.BaseCollection.on_snapshot` (signature only)
- `base_document.BaseDocumentRef.on_snapshot` (signature only)
- `base_query.BaseQuery.on_snapshot` (signature only)
- `collection.Collection.on_snapshot`
- `document.Document.on_snapshot`
- `query.Query.on_snapshot`
- `watch.Watch.__init__`
- `watch.Watch.for_document`
- `watch.Watch.for_query`

Closes #221

### CODE REVIEW

**Comment by tseaver** [comment]
Note to reviewers:  commit-by-commit is likely the most fruitful approach.

File: google/cloud/firestore_v1/async_document.py [hunk 1]
```diff
@@ -350,36 +369,39 @@ async def get(
...
-            request={"parent": self._document_path, "page_size": page_size},
-            metadata=self._client._rpc_metadata,
+            request=request, metadata=self._client._rpc_metadata, **kwargs,
         )
 
         while True:
             for i in iterator.collection_ids:
                 yield self.collection(i)
             if iterator.next_page_token:
+                next_request = request.cpoy()
```
Comment (tseaver) [inline google/cloud/firestore_v1/async_document.py]: Typo fixed in 9b4707a8124de962c6fbb00be79eb00b3954724a

File: google/cloud/firestore_v1/document.py [hunk 2]
```diff
@@ -415,25 +389,20 @@ def collections(
...
-            metadata=self._client._rpc_metadata,
-            **kwargs,
+            request=request, metadata=self._client._rpc_metadata, **kwargs,
         )
 
         while True:
             for i in iterator.collection_ids:
                 yield self.collection(i)
             if iterator.next_page_token:
+                next_request = request.cpoy()
```
Comment (tseaver) [inline google/cloud/firestore_v1/document.py]: Typo fixed in 1acbde389b497a4a87df7fa365f49f29dcca5084

**Review by tseaver — COMMENTED** [review]
Comments on commit-by-commit review reflecting later changes as the PR evolved.

File: google/cloud/firestore_v1/query.py [hunk 3]
```diff
@@ -125,9 +131,12 @@ def get(self, transaction=None) -> list:
             transaction
                 (Optional[:class:`~google.cloud.firestore_v1.transaction.Transaction`]):
                 An existing transaction that this query will run in.
-        If a ``transaction`` is used and it already has write operations
-        added, this method cannot be used (i.e. read-after-write is not
-        allowed).
+                If a ``transaction`` is used and it already has write operations
+                added, this method cannot be used (i.e. read-after-write is not
+                allowed).
+            retry (google.api_core.retry.Retry): Designation of what errors, if any,
```
Comment (crwilcox) [inline google/cloud/firestore_v1/query.py:137]: Is it possible to expose what default retries exist for a call? Or perhaps a way to modify this to indicate that if not set, there are still retries? That a user doesn't need to specify this information to get retry behavior?

Also, what would this need to be set to to have no retry? It is defaulted to None, would an empty list work?

File: tests/unit/v1/test_async_client.py [hunk 4]
```diff
@@ -220,10 +220,11 @@ def _next_page(self):
                     page, self._pages = self._pages[0], self._pages[1:]
                     return Page(self, page, self.item_to_value)
 
+        kwargs = _helpers.make_retry_timeout_kwargs(retry, timeout)
         iterator = _Iterator(pages=[collection_ids])
         firestore_api.list_collection_ids.return_value = iterator
 
-        collections = [c async for c in client.collections()]
+        collections = [c async for c in client.collections(**kwargs)]
```
Comment (crwilcox) [inline tests/unit/v1/test_async_client.py:227]: just confirming, you have added these changes to a few tests to get complete coverage? The changes of this PR shouldn't result in any required changes to users correct?

File: google/cloud/firestore_v1/query.py (hunk 3 above)
Comment (tseaver) [reply google/cloud/firestore_v1/query.py:137]: Exposing the actual default retries is infeasible:  the values are [buried in generated code](https://github.com/googleapis/python-firestore/blob/c122e4186808468a2ff82e9cc54b501809519859/google/cloud/firestore_v1/services/firestore/transports/base.py#L110-L300).  We could add language saying that if no value is passed, a system-supplied default will be used.

To allow turning off retries altogether, we would need to have the default be `google.api_core.gapic_v1.method.DEFAULT`, and then add it to the `**kwargs` only if it doesn't match that sentinel.

File: tests/unit/v1/test_async_client.py (hunk 4 above)
Comment (tseaver) [reply tests/unit/v1/test_async_client.py:227]: Right.  See `test_collections` and `test_collections_w_retry_timeout` below.

**Review by crwilcox — APPROVED** [review]

**Pull request merged**